SRGSSR/pillarbox-web
# feat: custom media data support and tracker control

## Description

Add support for custom media data when provided via the `mediaData` property to
the `player.src`. Allow `mediaComposition` properties to be added or
overwritten when loading a new source.

```javascript
// custom media data
player.src({
  src : 'urn:xyz',
  type: 'srgssr/urn',
  mediaData : {
    // custom properties
  }
})
```

Store the `DataProvider` in the player's `srgOptions`, allowing it to be changed
on the fly. This means it's no longer needed to extend the `SrgSsr` class to
redefine the default `DataProvider`. The only requirement is that the new
`DataProvider` implements the `getMediaCompositionByUrn` function and returns an
object of type `MediaComposition`.

```javascript
// custom data provider
player.options({
  srgOptions : {
    dataProvider : {
      getMediaCompositionByUrn(urn) {
        // ... body of the function
      }
    }
  }
})
```

Add support for the `trackers` property to the `player.options` in order to
store tracker instances. This makes it easier to access trackers directly from
the player. Use of this feature remains optional.

```javascript
// add custom tracker
const player = new Pillarbox('my-player', {
  trackers: {
    customTracker: new CustomTracker(player),
    anotherTracker: new AnotherTracker(player),
  }
})
```

Add the ability to disable trackers per playback session by introducing the
`disableTrackers` property provided directly to the `player.src`. As most trackers
take an instance of the player as a parameter. This will make it easier to access
the disableTrackers property to determine whether or not the tracker should be
disabled for the current playback session.

```javascript
// disable a tracker for the playback session
player.src({
  src : 'urn:xyz',
  type: 'srgssr/urn',
  disableTrackers : ['anotherTracker']
});

// disable all trackers for the playback session
player.src({
  src : 'urn:xyz',
  type: 'srgssr/urn',
  disableTrackers : true
})
```

Add support for SRG SSR-compliant media playback tracking.

## Changes made

- add `srgOptions` to Pillarbox default options
- rename `DataProviderService` to `DataProvider`
- move `DataProvider` from middleware functions parameter to its own function
- handle `srcOptions` enabling customized `mediaData` to be provided to the player
- add `srcOptions` to the `composeSrcMediaData` function parameters
- update `srgssr` unit tests
  - add `srgOptions` to the options in the mock player
  - rename `DataProviderService` to `DataProvider`
  - remove `DataProvider` from `middleware` function parameters
  - update some `middleware` test cases description
- add `trackers` to Pillarbox default options
- add the `SRGAnalytics` class and unit tests
- add `PlayerEvents` that contains all `mediaElement` events
- srgssr middleware
  - add a function to handle SRG SSR analytics
  - update `composeSrcMediaData` to handle `disableTrackers`
  - move the construction of the main resources array into its own function
  - update unit tests
- update `baseUrl` property to use IL 2.1
- handle disabled `titleBar`
- replace `bu` by `vendor`
- update `video.js` to version [8.5.3](https://github.com/videojs/video.js/releases/tag/v8.5.3)
- add script to check for `outdated` packages 
- update `eslint` allow unused vars prefixed by `_`

Resolves #32
